### PR TITLE
CoMeT 2019.12

### PR DESCRIPTION
**Pull request by nritsche**
- add archiver
- use redis to store data
- add locust stress test
- add request ID to logging

### CODE REVIEW

File: scripts/comet_archiver [hunk 1]
```diff
@@ -5,34 +5,33 @@ import logging
...
 parser.add_argument(
-        "-d", "--dump-path",
-        help="set dump path (default: '{}')".format(DEFAULT_DUMP_PATH),
-        default=DEFAULT_DUMP_PATH
-        )
-parser.add_argument(
-        "-i", "--scrape-interval",
-        help="set scrape interval in seconds (default: '{}')".format(DEFAULT_SCRAPE_INTERVAL),
-        default=DEFAULT_SCRAPE_INTERVAL,
+        "-t", "--failure-wait-time",
```
Comment (anjakefala) [inline scripts/comet_archiver]: parsed_args will convert this into a variable named `failure_wait_time`?

File: comet/archiver.py [hunk 2]
```diff
@@ -44,129 +64,157 @@ def __init__(self, data_dump_path, scrape_interval, broker_host, broker_port):
             )
             exit(1)
 
-        self.loop = asyncio.get_event_loop()
-        self.task = self.loop.create_task(self._scrape())
```
Comment (anjakefala) [inline comet/archiver.py]: Why did these get removed?

File: comet/archiver.py [hunk 3]
```diff
@@ -44,129 +64,157 @@ def __init__(self, data_dump_path, scrape_interval, broker_host, broker_port):
...
-                __version__, self.dir, datetime.timedelta(seconds=self.interval)
+        """Run comet archiver (forever)."""
+        logger.info("Started CoMeT Archiver {}.".format(__version__))
+
+        # names of the lists we are monitoring on redis
+        TYPES = ["archive_state", "archive_dataset"]
+
+        while True:
+            # Shuffle the list of types, otherwise the archiver can get stuck on a
+            # single invalid item in the first list, ignoring any items in the second
```
Comment (anjakefala) [inline comet/archiver.py]: D8

File: scripts/comet_archiver (hunk 1 above)
Comment (nritsche) [reply scripts/comet_archiver]: yes

File: comet/archiver.py (hunk 2 above)
Comment (nritsche) [reply comet/archiver.py]: The archiver got completely rewritten. The previous one read from file what it has to archive (and used asyncio). This one (doesn't and) reads its input from lists on redis.

File: .travis.yml [hunk 4]
```diff
@@ -48,17 +49,18 @@ before_script:
...
-    # Run pycodestyle on all .py files in all subfolders
+    # Run black and pydocstyle on all .py files in all subfolders
     - black --check .
     - find . -type d -o ! -name versioneer.py ! -name test_*.py ! -name _version.py -name \*.py -exec pydocstyle --convention=numpy --add-ignore=D105,D202 {} +
 
     # Run tests
     - export PYTHONPATH=.:$PYTHONPATH
     - cd tests
-    - pytest -s
+    - pytest -x
```
Comment (anjakefala) [inline .travis.yml]: Do you still want to keep this change to pytest?

Comment (nritsche) [reply .travis.yml]: Yes, why not?

**Comment by anjakefala** [comment]
Such an improvement!

**Pull request merged**